### PR TITLE
Explain how to create a custom pattern

### PR DESCRIPTION
**Pull request by petschki**
This explains how to create a custom mockup pattern in an addon with `plonecli`

See #1270 

----
📚 Documentation preview 📚: https://plone6--1748.org.readthedocs.build/

### CODE REVIEW

**Comment by petschki** [comment]
@stevepiercy I saw that you have updated the "get started" section recently but this doesn't have to do anything with mockup or patternslib at all IMHO ... this is more a general explanation of how to create an addon for the backend with plonecli and should be placed somewhere else (maybe new sub-chapter "Create Addon" in the backend chapter ?) ... for mockup the subtemplate `mockup_pattern` is interesting when creating a new one.

**Review by MrTango — DISMISSED** [review]
Thx for the fix. I was wondering if we could extract the bobtemplates/plonecli explanation into its own chapter.nthis way we can reference it in different places. Maybe even as an included block or something. I would like to avoid that we have this redundant infos getting outdated at some point. If we have the info in one place, it's easy to update.

**Comment by petschki** [comment]
Maybe we should do a main "development guide" chapter like the new admin guide here #1746 ...

- Development
- - create addon with cookiecutter
- - create addon with plonecli
- - create theme for Volto
- - create theme for Classic UI

**Comment by stevepiercy** [comment]
> @stevepiercy I saw that you have updated the "get started" section recently

Sorry, which "get started" section? I assume you mean https://6.docs.plone.org/install/index.html#get-started. If so, then should we rename [Create a project with Classic UI (stable release)](https://6.docs.plone.org/install/create-project-classic-ui.html) to, say, "Create a backend project"?

We need to rename them all anyway, because with the looming release of Volto 18 tomorrow, the whole stable/pre-release paradigm gets tossed out. Instead the two Volto items will be renamed with "Volto 17" and "Volto 18".

ICYMI, we're migrating toward the [Diataxis Framework](https://diataxis.fr/) for organization of documentation along two axes or four quadrants. Almost all of the current docs is a collection of how-to guides. "Backend" will be moved into separate guides, but mostly into the how-to guide, Admin Guide.

>  but this doesn't have to do anything with mockup or patternslib at all IMHO ... this is more a general explanation of how to create an addon for the backend with plonecli and should be placed somewhere else (maybe new sub-chapter "Create Addon" in the backend chapter ?) ... for mockup the subtemplate `mockup_pattern` is interesting when creating a new one.

Sorry, two "this"s and I am lost. I think your first "this" referred to [Create a project with Classic UI (stable release)](https://6.docs.plone.org/install/create-project-classic-ui.html), and the second "this" to this pull request [Mockup and Patternslib](https://plone6--1748.org.readthedocs.build/classic-ui/mockup.html). Assuming I guessed correctly, then first I will review this content, then we can determine where it should go, although I think @davisagli might have already figured it out for us before I could reply. 😁

**Comment by davisagli** [comment]
@stevepiercy No, @petschki was talking about the "get started" section of the file that's the focus of this PR (https://6.docs.plone.org/classic-ui/mockup.html#get-started). He's saying the general information about how to create an add-on using plonecli belongs in a separate section. I agree and I think the developer guide is a good place for that. (Can we merge the initial creation of the developer guide so that it's easier to work on new PRs that add things to it?)

> We need to rename them all anyway, because with the looming release of Volto 18 tomorrow, the whole stable/pre-release paradigm gets tossed out. Instead the two Volto items will be renamed with "Volto 17" and "Volto 18".

@stevepiercy I think I've already taken care of that in https://github.com/plone/documentation/pull/1746 in a way that will work after Volto 18 is released. Please take a look at that and give feedback before working on renaming them in a different way.

**Comment by petschki** [comment]
@stevepiercy sorry for the confusion and @davisagli thanks for clarifying 👍🏼 

I was talking about the changes here #1734

File: docs/classic-ui/mockup.md [hunk 1]
```diff
@@ -52,6 +52,80 @@ You can check the full list of available features using the `-l` parameter:
...
+|   |   |   |   |   |   ├── addon-remote.min.js
+|   |   |   |   |   |   ├── addon-remote.min.js.map
+|   |   |   |   |   |   ├── addon.min.js
+|   |   |   |   |   |   ├── addon.min.js.map
+```
+
+Note that `plonecli` also creates an XML file in {file}`src/project/addon/profiles/default/registry/bundles.xml`, which registers the {file}`addon-remote.min.js` file in the resource registry.
+
+```{important}
+You must re-import your profile with an upgrade step if you installed your add-on in Plone before adding the pattern.
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md:125]: How does one do this step?

File: docs/classic-ui/mockup.md [hunk 2]
```diff
@@ -52,6 +52,80 @@ You can check the full list of available features using the `-l` parameter:
...
+```
+
+You can test your pattern now with the browser view `@@addon-pattern-demo` (see {file}`/src/project/addon/browser/pattern-demo.pt`)
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]: Can you explain further? Do I load something in a web browser? What does the file have to do with testing?

File: docs/classic-ui/mockup.md [hunk 3]
```diff
@@ -52,6 +52,80 @@ You can check the full list of available features using the `-l` parameter:
...
+Alternatively you can implement it in your own templates by adding the CSS class `pat-testpattern` to a tag.
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]: Does this mean, for example, `<some_html_tag class="pat-testpattern">`?

**Review by stevepiercy — CHANGES_REQUESTED** [review]
This is awesome! Thanks for the contribution. My review is my pushed commits, with a couple of noobie questions.

File: docs/classic-ui/mockup.md (hunk 1 above)
Comment (petschki) [reply docs/classic-ui/mockup.md:125]: Simplest way: uninstall/install the addon

I searched for the genericsetup documentation and got a huge old document here https://5.docs.plone.org/develop/addons/components/genericsetup.html#upgrade-steps which is still 85% valid ... would be goot to polish it and take it over to the actual docs.

File: docs/classic-ui/mockup.md (hunk 2 above)
Comment (petschki) [reply docs/classic-ui/mockup.md]: ah ok ... its not really a test but a demo setup how to integrate your patter in a template.
This is accessible through the browser with `http://localhost:8080/Plone/@@addon-pattern-demo` depending on the development port setup and the Plone Site ID.

Not sure if we should mention it at all but instead explain how its done with css class (see below)

File: docs/classic-ui/mockup.md (hunk 3 above)
Comment (petschki) [reply docs/classic-ui/mockup.md]: yes exactly ... maybe better explain the usage here instead of mentioning the demo browser view.

Comment (stevepiercy) [reply docs/classic-ui/mockup.md]: Done in 2e0e275897d4f6e6881c2d6c13ce721322be33b4

File: docs/classic-ui/mockup.md (hunk 1 above)
Comment (stevepiercy) [reply docs/classic-ui/mockup.md:125]: Yeah, GenericSetup docs have not been ported to Plone 6 yet. I don't know which parts are valid or not. We can do a separate PR for them. I can convert this file https://github.com/plone/documentation/blob/5.2/develop/addons/components/genericsetup.rst from .rst to .md, then you can fiddle with it. Where should its content go? I think most of it is how-to guide.

File: docs/classic-ui/mockup.md (hunk 2 above)
Comment (stevepiercy) [reply docs/classic-ui/mockup.md]: I'll go with what you think is best. We don't need to document all usage, but the most useful to developers. I added an example of the class, and if that is "best", then let's go with it.

File: docs/classic-ui/mockup.md (hunk 1 above)
Comment (petschki) [reply docs/classic-ui/mockup.md:125]: That would be great. Best place would be inside a "howto create a backend addon" guide

File: docs/classic-ui/mockup.md (hunk 2 above)
Comment (petschki) [reply docs/classic-ui/mockup.md]: I updated the info for the demo browser view. see https://github.com/plone/documentation/pull/1748/commits/b25cf80328a3fbcc911d702a7d2195a4a2858456 ...  lets keep it that way.

File: docs/classic-ui/mockup.md [hunk 4]
```diff
@@ -52,6 +52,86 @@ You can check the full list of available features using the `-l` parameter:
...
+yarn install
+```
+
+When you finish writing your JavaScript code, you have to build the bundle with the following command.
+
+```shell
+yarn build
+```
+
+This creates the webpack chunks, the JS bundle files, a demo browser view in your addon package:
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]:
```suggestion
This creates the webpack chunks, the JavaScript bundle files, and a demo browser view in your add-on package.
```

File: docs/classic-ui/mockup.md [hunk 5]
```diff
@@ -52,6 +52,86 @@ You can check the full list of available features using the `-l` parameter:
 plonecli -l
 ```
 
+## Create a custom pattern
+
+To create a custom {term}`pattern` in your add-on, use the `mockup_pattern` {term}`bobtemplate`:
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]:
```suggestion
To create a custom {term}`pattern` in your add-on, use the `mockup_pattern` {term}`bobtemplate`.
```

File: docs/classic-ui/mockup.md [hunk 6]
```diff
@@ -52,6 +52,86 @@ You can check the full list of available features using the `-l` parameter:
...
+|   |   |   |   |   ├── bundles
+|   |   |   |   |   |   ├── chunks
+|   |   |   |   |   |   ├── addon-remote.min.js
+|   |   |   |   |   |   ├── addon-remote.min.js.map
+|   |   |   |   |   |   ├── addon.min.js
+|   |   |   |   |   |   ├── addon.min.js.map
+|   |   |   |   ├── pattern-demo.pt
+```
+
+There is also a XML file in `src/project/addon/profiles/default/registry/bundles.xml` which registers the `addon-remote.min.js` in the resources registry.
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]:
```suggestion
There is also an XML file in {file}`src/project/addon/profiles/default/registry/bundles.xml` which registers the {file}`addon-remote.min.js` in the resources registry.
```

File: docs/classic-ui/mockup.md [hunk 7]
```diff
@@ -52,6 +52,86 @@ You can check the full list of available features using the `-l` parameter:
...
+
+```{important}
+You must re-import your profile with an upgrade step if you installed your add-on in Plone before adding the pattern by uninstall/install the addon in the controlpanel or writing a genericsetup upgrade step.
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]:
```suggestion
You must re-import your profile with an upgrade step if you installed your add-on in Plone before adding the pattern.
Uninstall, then re-install, the add-on in the control panel.
Alternatively you can write a GenericSetup upgrade step.
```

File: docs/classic-ui/mockup.md [hunk 8]
```diff
@@ -52,6 +52,86 @@ You can check the full list of available features using the `-l` parameter:
...
+```
+
+You can access the demo browser view in your browser with `http://localhost:8080/Plone/@@addon-pattern-demo`
```
Comment (stevepiercy) [inline docs/classic-ui/mockup.md]:
```suggestion
You can access the demo browser view in your browser with `http://localhost:8080/Plone/@@addon-pattern-demo`.
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Minor grammar tweaks. Otherwise LGTM! Thank you!

**Review by stevepiercy — APPROVED** [review]

File: docs/classic-ui/mockup.md (hunk 1 above)
Comment (stevepiercy) [reply docs/classic-ui/mockup.md:125]: I created a new issue to track this at https://github.com/plone/documentation/issues/1754

**Review by MrTango — APPROVED** [review]

File: docs/classic-ui/mockup.md (hunk 1 above)
Comment (stevepiercy) [reply docs/classic-ui/mockup.md:125]: PR started in https://github.com/plone/documentation/pull/1757

**Pull request merged**